### DISCUSSION TITLE
C# / JS types

### DISCUSSION THREAD

**Discussion by KoalaBear84** [sebastienros/jint#1068] [Q&A]
As I cannot reopen a ticket, a 'successor' to #1064

I'm having issues when using C# Math, instead of a difficult way to Round in JavaScript but cannot get it to work. I guess this might be because of matching C# and JS types?

```csharp
public static class MathExtensions
{
	public static double Round(this double value, int decimals) => Math.Round(value, decimals);
}

Engine engine = new Engine(options =>
{
    options
      .AllowClr()
      .AddExtensionMethods(typeof(MathExtensions));
});

double result = engine
	// Returns 1 instead of 1.2
	//.Evaluate("System.Math.Round(parseFloat('1.2345678'), 2)")

	// Returns 1 instead of 1.2
	//.Evaluate("System.Math.Round(new Number(1.23456), 2)")

	// Works
	.Evaluate("new Number(1.23456).round(2)")

	.AsNumber();

Console.WriteLine(result.ToString());
```

**Update:**
It indeed works the extension method way. Why doesn't the System.Math.Round work, and returns an 'integer'? (guess)

**Comment by lahma** [comment]
> As I cannot reopen a ticket, a 'successor' to #1064

You don't need to have an open ticket to add comments, and please use discussions for questions.

This works just fine:  `Assert.Equal(1.23, new Engine(o => o.AllowClr()).Evaluate("System.Math.Round(parseFloat('1.2345678'), 2)").AsNumber());
`

**Reply by KoalaBear84** [reply]
Hmm, here it really doesn't work, really returning "1", with both a dot and a comma. Is it dependent on any localization? If it should be it should at least work in any of these cases.

![image](https://user-images.githubusercontent.com/5886900/151491970-e04a5a7f-908a-4cc6-b6eb-6fc52a2efb41.png)

![image](https://user-images.githubusercontent.com/5886900/151491917-e29ebdef-6a31-40f6-a4da-b017a15e7bc6.png)

![image](https://user-images.githubusercontent.com/5886900/151492418-3cb55ec3-b63f-473e-ab01-0e952f7efe3f.png)

**Reply by lahma** [reply]
As I don't have your computer to test with, could you write a failing unit test case against the latest main branch here?

**Reply by KoalaBear84** [reply]
OK, I'll try to get to that today.

**Reply by KoalaBear84** [reply]
I can only reproduce it in .NET Core, looks like it works in .NET Framework. But shouldn't it fail also in the tests then? Or is the code not really running 'under' .NET Core?

If I only do "parseFloat('1.2345678')" it does return the same for both.

![image](https://user-images.githubusercontent.com/5886900/151498556-ab3c7e21-6103-447b-aae6-56254b20c303.png)

Only changed the following:

```csharp
public NumberTests()
{
	_engine = new Engine(o => o.AllowClr())
			.SetValue("log", new Action<object>(Console.WriteLine))
			.SetValue("assert", new Action<bool>(Assert.True))
			.SetValue("equal", new Action<object, object>(Assert.Equal));
}

[Theory]
[InlineData(2, 1.23d)]
[InlineData(4, 1.2346d)]
public void SystemMathParseFloat(int digits, double result)
{
	var value = _engine.Evaluate($"System.Math.Round(parseFloat('1.2345678'), {digits})").AsNumber();
	Assert.Equal(result, value);
}
```

Clean project you can open:
[TestJintMathRound.zip](https://github.com/sebastienros/jint/files/7956315/TestJintMathRound.zip)

**Reply by lahma** [reply]
NET6 in your runner means that it's running on new .NET, not full framework.
